##                   <font color='blue'>        Estudo de Viabilidade de jogos Online na Fun Corp  🎮 </font> 


Este relatório contém o código usado para o estudo de viabilidade de jogos online na Fun Corp, foram criados módulos em um arquio complementar para que o código seja apresentado de forma mais clara e menos complexa. Os módulos foram criados separadamente para melhor manutenção do código posteriormente.

### Leitura do Arquivo 

In [1]:
# Importando do módulo csv_reader a classe CSVLoader.
from csv_reader import CSVLoader

# Criando uma instância da classe CSVLoader, passando o caminho do arquivo CSV.
loader = CSVLoader('steam_games.csv')

# Chamando o método load_csv para carregar o arquivo CSV.
headers, data = loader.load_csv()

# Exibindo o cabeçalho como teste.
print("Cabeçalhos:", headers)

Cabeçalhos: ['AppID', 'Name', 'Release date', 'Estimated owners', 'Peak CCU', 'Required age', 'Price', 'DLC count', 'About the game', 'Supported languages', 'Full audio languages', 'Reviews', 'Header image', 'Website', 'Support url', 'Support email', 'Windows', 'Mac', 'Linux', 'Metacritic score', 'Metacritic url', 'User score', 'Positive', 'Negative', 'Score rank', 'Achievements', 'Recommendations', 'Notes', 'Average playtime forever', 'Average playtime two weeks', 'Median playtime forever', 'Median playtime two weeks', 'Developers', 'Publishers', 'Categories', 'Genres', 'Tags', 'Screenshots', 'Movies']


### Transformação em um dicionário

In [2]:
# Importando a classe do módulo que irá transformar os dados em um dicionário.
from data_dicts import CSVtoDict

# Criando uma instância da classe CSVLoader, passando o caminho do arquivo CSV.
csv_loader = CSVtoDict('steam_games.csv')

# Chamando o método load_csv_to_dict para carregar o arquivo CSV como um dicionário.
data_dicts = csv_loader.load_csv_to_dict()

# Exibindo os dados transformados em um dicionário, correlacionando o cabeçalho às informações 
# das demais linhas. 
for data_dict in data_dicts[:2]: # Limitando a exibição por conta da extensão do arquivo.
    for key, value in data_dict.items():
        print(f"{key}: {value}") # Certificando que o código foi bem executado.
    print('---') 


AppID: 1967950
Name: Moi Mei: Hidden Objects
Release date: May 18, 2022
Estimated owners: 0 - 20000
Peak CCU: 1
Required age: 0
Price: 9.99
DLC count: 0
About the game: The honeymoon in May went clearly not the way a self-respecting honeymoon should go. As soon as they arrived on the paradise island lost in the Caribbean, Mae and Jason immediately got into serious trouble — the local voodoo priestess Mama Agatha fell in love with Jason. The sorceress, realizing that she would not be able to recapture her husband from Mae so easily, resorted to the help of witchcraft and turned Jason into a natural nzambi! And now the Snake, in order to save his beloved, will have to master all the subtleties of witchcraft, enlist the support of cunning and greedy Voodoo spirits and challenge the insidious Mother Agatha! Do not pass by, we have here: a strong and determined heroine a whole family of spirits to be appeased magical atmosphere ritual drums
Supported languages: ['English', 'Russian']
Full a

## 🔎 Respondendo as Perguntas de Negócio após Manipulação dos Dados  

### Pergunta 1 🎮
Qual o percentual de jogos gratuitos e pagos na plataforma? 

In [3]:
# Importando a classe do módulo que irá verificar o percentual. 
from percentage import GameStats

game_stats = GameStats(data_dicts)
game_stats.calculate_free_paid_percentage()

Porcentagem de jogos gratuitos: 5.00%
Porcentagem de jogos pagos: 95.00%


#### Resposta
A análise dos dados da Steam revelou que 95% dos jogos são pagos, indicando uma preferência do consumidor por conteúdo de qualidade e a necessidade de estratégias de monetização eficazes. Para a Fun Corp., isto sugere uma oportunidade de entrar no mercado de jogos digitais focando em títulos pagos que ofereçam experiências únicas e envolventes, destacando-se pela qualidade e originalidade. Identificar nichos e preferências dos jogadores será crucial para o sucesso neste segmento competitivo.


### Pergunta 2 🎮
Qual o ano com o maior número de novos jogos? (Em caso de empate irá retornar uma lista com os anos empatados). 

In [4]:
# Importando a classe do módulo para verificar registros de jogos por ano.
from common_years import find_most_common_years

most_common_years = find_most_common_years(data_dicts)

# Verificar se há empate e, se houver, imprimi todos os anos com o maior número de lançamentos.
if len(most_common_years) > 1:
    print("Houve um empate entre os seguintes anos com o maior número de novos jogos:")
    for year in most_common_years:
        print(year)
else:
    print("Ano com o maior número de novos jogos:", most_common_years[0])

Ano com o maior número de novos jogos: 2018


#### Resposta 
A análise dos dados da Steam destaca 2018 como o ano com maior lançamento de jogos, indicando uma época de alta inovação e demanda no mercado digital. Para a Fun Corp., compreender as condições de sucesso desse ano e a dinâmica de mercado sem empates em lançamentos nos outros anos é crucial para planejar estrategicamente sua entrada e posicionamento no mercado de jogos digitais.

### Pergunta 3 🎮
Quais são os TOP 3 gêneros de jogos mais comuns lançados nos anos com o maior número de novos jogos?

In [5]:
# O método Counter serva para percorrer os dados e contabilizá-los.
from collections import Counter

# Função para identificar os gêneros mais comuns nos anos com o maior número de novos jogos.
def most_common_genres_in_top_years(data_dicts):
    """
    >>> data_dicts = [{"Release date": "Sep 13, 2018", "Genres": "Casual,Indie"},]
    >>> most_common_genres_in_top_years(data_dicts)
    {2018: {'total_games': 1, 'top3_genres': [('Casual,Indie', 1)]}}
    """
    
# Identificar os anos com o maior número de novos jogos.
    most_common_years = find_most_common_years(data_dicts)
    
# Dicionário para armazenar a frequência dos gêneros nos anos mais comuns.
    genres_frequency = Counter()
    
# Dicionário para armazenar os top 3 gêneros mais comuns por ano.
    top3_genres_by_year = {}
    
# Iterando sobre os dados para contar a frequência dos gêneros em cada ano.
    for game in data_dicts:
        release_year = int(game["Release date"].split()[-1])
        if release_year in most_common_years:
            genres = game["Genres"].split(", ")  # Separar os gêneros por vírgula.
            # Verificar se o campo de gênero está vazio.
            if genres != ['']:  # Se não estiver vazio, atualizar a contagem de gêneros.
                genres_frequency.update(genres)
    
# Adicionar os top 3 gêneros mais comuns por ano ao dicionário.
    for year in most_common_years:
        total_games_in_year = 0  # Inicializar o total de jogos no ano.
        genres = []
        for game in data_dicts:
            release_year = int(game["Release date"].split()[-1])
            if release_year == year:
                total_games_in_year += 1  # Atualizar o total de jogos no ano.
                game_genres = game["Genres"].split(", ")
                if game_genres != ['']:  # Validação para caso a variável esteja vazia.
                    genres.extend(game_genres)
        top3_genres_by_year[year] = {"total_games": total_games_in_year,
                                     "top3_genres": Counter(genres).most_common(3)}
    
# Retornar os top 3 gêneros mais comuns nos anos com o maior número de novos jogos.
    return top3_genres_by_year


In [6]:
# Chamando a função com os dados da variável 'data_dicts' onde foram convertidos em dicionário.
top3_genres_by_year = most_common_genres_in_top_years(data_dicts)

# Imprimindo os resultados.
print("Top 3 gêneros mais comuns nos anos com o maior número de novos jogos")
for year, genres_info in top3_genres_by_year.items():
    print(f"Ano: {year} - Total de jogos lançados: {genres_info['total_games']}")
    for rank, (genre, frequency) in enumerate(genres_info['top3_genres'], start=1):
        print(f"  {rank}. {genre}: {frequency} jogos")
    print()

Top 3 gêneros mais comuns nos anos com o maior número de novos jogos
Ano: 2018 - Total de jogos lançados: 5
  1. Casual,Indie: 1 jogos
  2. Action,Adventure,Indie: 1 jogos
  3. Adventure,Indie: 1 jogos



#### Resposta 
Em nossa análise focalizada no ano de 2018, identificamos que este ano não só teve o maior número de lançamentos na Steam, mas também revelou preferências específicas de gênero. Com um total de 5 jogos lançados, os gêneros mais comuns foram:

Casual, Indie: 1 jogo
Action, Adventure, Indie: 1 jogo
Adventure, Indie: 1 jogo
Esta distribuição destaca a popularidade dos gêneros Indie, juntamente com uma combinação de Casual, Action e Adventure, refletindo a diversidade e o interesse variado dos consumidores. Para a Fun Corp., esses insights são valiosos para direcionar o desenvolvimento de futuros jogos digitais, alinhando-se às tendências de gênero mais procuradas.

## Criando a Amostra para Teste 🎯
Para realizar o teste foi separada uma amostra contendo dados aleatórios de 20 jogos, depois a amostra foi salva em um outro arquivo csv. 

In [7]:
# Importando a biblioteca random para gerar os dados aleatoriamente. 
import random

# O método seed irá garantir que sempre que for chamada, a amostra continua com os mesmos dados.
random.seed(42)

# Criando a amostra usando os dados do dicionário criado. 
sample_20games = random.sample(data_dicts, 20)

#print(sample_20games[:1]) 
# Descomentar caso queira observar um item da amostra.

In [8]:
# Confirmando quantidade de elementos na amostra. 
sample= len(sample_20games)
print(sample)

20


## Salvando a Amostra

In [ ]:
#Salvando a amostra em um arquivo CSV
import csv
# Caminho do arquivo CSV onde você deseja salvar os dados
arquivo_csv = 'amostra_de_20jogos.csv'

# Obtendo os nomes das chaves (cabeçalho) dos dicionários
headers = sample_20games[0].keys()

# Escrevendo os dados no arquivo CSV
with open(arquivo_csv, 'w', newline='', encoding='utf-8') as file:
    writer = csv.DictWriter(file, fieldnames=headers)
    
    # Escreve o cabeçalho no arquivo
    writer.writeheader()
    
    # Escreve os dados dos jogos no arquivo
    for jogo in sample_20games:
        writer.writerow(jogo)

print("Amostra de 20 jogos salva com sucesso em", arquivo_csv)

## Manipulando a Amostra Teste 🎯

In [10]:
# Transformando em dicionário 
# Importando a classe do módulo que irá transformar os dados em um dicionário.
from data_dicts import CSVtoDict

# Criando uma instância da classe CSVLoader, passando o caminho do arquivo CSV.
csv_loader = CSVtoDict('amostra_de_20jogos.csv')

# Chamando o método load_csv_to_dict para carregar o arquivo CSV como um dicionário.
amostra_dicts = csv_loader.load_csv_to_dict()

# Exibindo os dados transformados em um dicionário, correlacionando o cabeçalho às informações 
# das demais linhas. 
for data_dict in amostra_dicts[:2]: # Limitando a exibição por conta da extensão do arquivo.
    for key, value in data_dict.items():
        print(f"{key}: {value}") # Certificando que o código foi bem executado.
    print('---') 


AppID: 675010
Name: MudRunner
Release date: Oct 30, 2017
Estimated owners: 1000000 - 2000000
Peak CCU: 559
Required age: 0
Price: 19.99
DLC count: 4
About the game: MudRunner is the ultimate off-road experience, putting the players in the driver seat and daring them to take charge of incredible all-terrain vehicles, venturing across extreme Siberian landscapes with only a map and compass as guides! Drive 19 powerful all-terrain vehicles, each with its own characteristics and attachable equipment. Complete your objectives and deliveries by enduring perilous conditions across wild, untamed landscapes in extreme conditions with dynamic day-night cycles. Explore an immersive sandbox environment, enhanced by improved graphics. Overcome muddy terrain, raging rivers and other obstacles that all realistically react to the weight and movement of your vehicle powered by the game's advanced physics engine. With your map, compass, winch, and your driving skills as allies, go solo or join up to thr

### Respondendo as Perguntas de Negócio usando a biblioteca doctest para teste.

Para o teste foram usados os códigos completos que estavam como módulos anteriormente, para visualização completa.

### Pergunta 1
Qual o percentual de jogos gratuitos e pagos na plataforma?

In [11]:
class GameStats:
    def __init__(self, game_data):
        self.game_data = game_data

    def calculate_free_paid_percentage(self):
        """
        Bloco para o doctest
        >>> game_data = [{'Name': 'Jogo 1', 'Price': '0.0'}]
        
        >>> game_stats = GameStats(game_data)
        >>> game_stats.calculate_free_paid_percentage()
        Porcentagem de jogos gratuitos: 100.00%
        Porcentagem de jogos pagos: 0.00%
        """
        
        free_count = 0
        paid_count = 0
        
        # Iteração pelos dicionários para contar jogos gratuitos e pagos.
        for game in self.game_data:
            price = game['Price'] # Variável onde se encontra a informação sobre preços.
            if price == '0.0' or price == '[0.0]':  # Considerando variações na representação de preço gratuito.
                free_count += 1
            else:
                paid_count += 1
                
        total_games = free_count + paid_count
        if total_games > 0:  # Evitando divisão por zero.
            free_percentage = (free_count / total_games) * 100 # Transformação em porcentagem. 
            paid_percentage = (paid_count / total_games) * 100
            print(f"Porcentagem de jogos gratuitos: {free_percentage:.2f}%")
            print(f"Porcentagem de jogos pagos: {paid_percentage:.2f}%")
        else:
            print("Não há jogos para analisar.")
            
game_stats = GameStats(amostra_dicts)
game_stats.calculate_free_paid_percentage()

if __name__ == "__main__":
    import doctest
    doctest.testmod(verbose=True)

Porcentagem de jogos gratuitos: 5.00%
Porcentagem de jogos pagos: 95.00%
Trying:
    game_data = [{'Name': 'Jogo 1', 'Price': '0.0'}]
Expecting nothing
ok
Trying:
    game_stats = GameStats(game_data)
Expecting nothing
ok
Trying:
    game_stats.calculate_free_paid_percentage()
Expecting:
    Porcentagem de jogos gratuitos: 100.00%
    Porcentagem de jogos pagos: 0.00%
ok
Trying:
    data_dicts = [{"Release date": "Sep 13, 2018", "Genres": "Casual,Indie"},]
Expecting nothing
ok
Trying:
    most_common_genres_in_top_years(data_dicts)
Expecting:
    {2018: {'total_games': 1, 'top3_genres': [('Casual,Indie', 1)]}}
ok
3 items had no tests:
    __main__
    __main__.GameStats
    __main__.GameStats.__init__
2 items passed all tests:
   3 tests in __main__.GameStats.calculate_free_paid_percentage
   2 tests in __main__.most_common_genres_in_top_years
5 tests in 5 items.
5 passed and 0 failed.
Test passed.


#### Resposta manual para comparação

Preço //	Frequência de Jogos comprados com esse preço

9.99 //     4

0.99 //    	4

19.99 //    2

2.99 //    	2

4.99 //    	2

17.99 //   	1

0.00 //	    1

10.49 //    1

6.99 //    	1

2.39 //    	1

3.99 //    	1

Total de jogos = 4 + 4 + 2 + 2 + 2 + 1 + 1 + 1 + 1 + 1 + 1 = 20 jogos
Número de jogos gratuitos: 1
Porcentagem de jogos gratuitos= (1/20)×100=5% 
Número de jogos pagos: 19
Porcentagem de jogos pagos=(19/20)×100=95%

Assim, 5% dos jogos são gratuitos e 95% são pagos



### Pergunta 2

In [12]:
from collections import Counter
# A classe Counter do módulo collections serve para contar uma ocorrência em um conjunto de dados.

def find_most_common_years(data_dicts):
    """
    Bloco para o doctest
    >>> data_dicts = [{'Name': 'Jogo 1', 'Release date': 'January 1, 2021'}]
    >>> find_most_common_years(data_dicts)
    [2021]
    """
    # Extrai os anos de lançamento dos jogos e conta quantos jogos foram lançados em cada ano.
    release_years = Counter(int(game["Release date"].split()[-1]) for game in data_dicts)

    # Encontra o ano (ou anos) com o maior número de lançamentos.
    max_count = max(release_years.values())
    most_common_years = [year for year, count in release_years.items() if count == max_count]

    return most_common_years

most_common_years = find_most_common_years(data_dicts)

# Verificar se há empate e, se houver, imprimi todos os anos com o maior número de lançamentos.
if len(most_common_years) > 1:
    print("Houve um empate entre os seguintes anos com o maior número de novos jogos:")
    for year in most_common_years:
        print(year)
else:
    print("Ano com o maior número de novos jogos:", most_common_years[0])
    
if __name__ == "__main__":
    import doctest
    doctest.testmod(verbose=True)


Ano com o maior número de novos jogos: 2018
Trying:
    game_data = [{'Name': 'Jogo 1', 'Price': '0.0'}]
Expecting nothing
ok
Trying:
    game_stats = GameStats(game_data)
Expecting nothing
ok
Trying:
    game_stats.calculate_free_paid_percentage()
Expecting:
    Porcentagem de jogos gratuitos: 100.00%
    Porcentagem de jogos pagos: 0.00%
ok
Trying:
    data_dicts = [{'Name': 'Jogo 1', 'Release date': 'January 1, 2021'}]
Expecting nothing
ok
Trying:
    find_most_common_years(data_dicts)
Expecting:
    [2021]
ok
Trying:
    data_dicts = [{"Release date": "Sep 13, 2018", "Genres": "Casual,Indie"},]
Expecting nothing
ok
Trying:
    most_common_genres_in_top_years(data_dicts)
Expecting:
    {2018: {'total_games': 1, 'top3_genres': [('Casual,Indie', 1)]}}
ok
3 items had no tests:
    __main__
    __main__.GameStats
    __main__.GameStats.__init__
3 items passed all tests:
   3 tests in __main__.GameStats.calculate_free_paid_percentage
   2 tests in __main__.find_most_common_years
   2 tes

#### Resposta manual para comparação


Data	// Nº Novos Jogos por Data


Oct 30, 2017   // 	1

May 18, 2022   // 	1

Oct 4, 2018   //  	1

Sep 13, 2018   // 	1

Aug 30, 2017    //	1

Sep 27, 2021    //	1

Nov 26, 2018    //	1

Nov 13, 2022    //	1

Mar 31, 2023    //	1

Dec 8, 2020     //	1

Mar 13, 2018    //	1

Dec 1, 2020     //	1

Jan 15, 2022    //	1

Sep 17, 2018    //	1

Mar 31, 2017    //	1

Dec 10, 2012    //	1

Jan 11, 2022    //	1

Mar 28, 2023    //	1

Dec 13, 2019    //	1

Jun 17, 2020    //	1

O ano de 2018 contabilizando como o ano de maior lançamento de novos jogos: 5.


##### Os testes demonstraram que o código está correto e a resposta são correspondentes ao esperado.